circle/src/domain.rs

```diff
@@ -0,0 +1,359 @@
+            // is_transition: v_0(self.shift.rotate(p)),
+            is_transition: self.shift.rotate(p).real() - Ext::one(),
+            inv_zeroifier: zeroifier.inverse(),
+        }
+    }
+
+    // wow, really slow!
+    #[instrument(skip_all, fields(log_n = %coset.log_n))]
+    fn selectors_on_coset(&self, coset: Self) -> LagrangeSelectors<Vec<Self::Val>> {
+        let sels = coset
+            .points()
+            .map(|p| self.selectors_at_point(point_to_univariate(p).unwrap()))
+            .collect_vec();
+        LagrangeSelectors {
+            is_first_row: sels.iter().map(|s| s.is_first_row).collect(),
+            is_last_row: sels.iter().map(|s| s.is_last_row).collect(),
+            is_transition: sels.iter().map(|s| s.is_transition).collect(),
+            inv_zeroifier: sels.iter().map(|s| s.inv_zeroifier).collect(),
+        }
+    }
+
+    fn split_domains(&self, num_chunks: usize) -> Vec<Self> {
+        assert!(self.is_standard());
+        let log_chunks = log2_strict_usize(num_chunks);
+        self.points()
+            .take(num_chunks)
+            .map(|shift| CircleDomain {
+                log_n: self.log_n - log_chunks,
+                shift,
+            })
+            .collect()
+    }
+
+    /*
+    chunks=2:
+
+          1 . 1
+         .     .
+        0       0 <-- start
+        .   .   - (1,0)
+        0       0
+         .     .
+          1 . 1
+
+
+    idx -> which chunk to put it in:
+    chunks=2: 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0
+    chunks=4: 0 1 2 3 3 2 1 0 0 1 2 3 3 2 1 0
+    */
+    fn split_evals(
+        &self,
+        num_chunks: usize,
+        evals: RowMajorMatrix<Self::Val>,
+    ) -> Vec<RowMajorMatrix<Self::Val>> {
+        let log_chunks = log2_strict_usize(num_chunks);
+        assert!(evals.height() >> (log_chunks + 1) >= 1);
+        let width = evals.width();
+        let mut values: Vec<Vec<Self::Val>> = vec![vec![]; num_chunks];
+        let mut rows = evals.rows();
+        for _ in 0..(evals.height() >> (log_chunks + 1)) {
+            for chunk in values.iter_mut() {
+                chunk.extend_from_slice(rows.next().unwrap());
+            }
+            for chunk in values.iter_mut().rev() {
+                chunk.extend_from_slice(rows.next().unwrap());
+            }
+        }
+        assert!(rows.next().is_none());
+
+        values
+            .into_iter()
+            .map(|v| RowMajorMatrix::new(v, width))
+            .collect()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use std::collections::HashSet;
+
+    use itertools::izip;
+    use p3_mersenne_31::Mersenne31;
+    use rand::{thread_rng, Rng};
+
+    use super::*;
+    use crate::util::{eval_circle_polys, gemv_tr};
+    use crate::Cfft;
+
+    fn assert_is_twin_coset<F: ComplexExtendable>(d: CircleDomain<F>) {
+        let pts = d.points().collect_vec();
+        let half_n = pts.len() >> 1;
+        for (l, r) in izip!(&pts[..half_n], pts[half_n..].iter().rev()) {
```

Comment:
nit: with two generators we might as well use the `zip` method directly

Reply:
Something in my brain prefers the symmetry of `izip!`. It's almost always shorter and cleaner (`&xs`/`&mut xs` vs `xs.iter()`/`xs.iter_mut()`) and very occasionally has performance benefits (notably for multizips) so we might as well default to using it everywhere.

That said, it's a style thing so I defer to the taste of @dlubarov

Reply:
I guess the "something in my brain" is just years of python

Reply:
I like Python, we can keep it as is, if there is no overhead in having a compatible multiple zip iterator :)